Add User.mfa_required?

kevinlinxc opened the pull request:
Adds a function to check if a User requires mfa due to the number of downloads on one of their gems in preparation for future MFA requirement work.

Edit, more info:

The mfa_required function will check if a user needs to enable mfa because one of the packages they own passed the MFA-required downloads threshold. Eventually, this will block them from taking some actions that require authentication.


Doesn't use the function/block the user's access to anything.

jenshenny commented on app/models/concerns/user_multifactor_methods.rb:
```diff
@@ -61,11 +61,16 @@ def strong_mfa_level?
 
+    def mfa_required?
+      return false if strong_mfa_level?
+      rubygems.mfa_required_from_downloads.any?
```
super nit: add a newline to match the style of `mfa_recommended?` above.

jenshenny reviewed (approved):
LGTM! Since `mfa_required?` is a private method, we should add a public interface for us to use (ie. required versions of `mfa_recommended_not_yet_enabled?`, `mfa_recommended_weak_level_enabled?`) either in this PR or a fast followup.

kevinlinxc commented:
Added mfa_required_not_yet_enabled? and mfa_required_weak_level_enabled? and tests. The weak level checks will be deleted after UI Only is removed, right?

simi commented:
:thinking: It is hard to do review when those methods are not used yet. Are they going to be part of some next PR as well?

bettymakes commented on app/models/rubygem.rb:
```diff
@@ -33,9 +33,12 @@ class Rubygem < ApplicationRecord
 
+  # TODO: Change MFA_RECOMMENDED_THRESHOLD to 150_000_000 and MFA_REQUIRED_THRESHOLD to 165_000_000
   MFA_RECOMMENDED_THRESHOLD = 165_000_000
+  MFA_REQUIRED_THRESHOLD = 180_000_000
```
Double checking here: @jenshenny, you're implying `MFA_REQUIRED_THRESHOLD` is arbitrary for now because we're [in the midst of realigning](https://github.com/rubygems/rubygems.org/pull/3089#discussion_r898305452) on whether the thresholds are 150M/165M vs. 165M/180M. Is that correct?

jenshenny replied:
Mostly yeah. Since we are deciding between 165/180M downloads for required, 180 was chosen for now because `MFA_REQUIRED_THRESHOLD` can't be set to 165M because `MFA_RECOMMENDED_THRESHOLD` has to be strictly less than the `MFA_REQUIRED_THRESHOLD` (tests will break, etc).

bettymakes commented on app/models/rubygem.rb:
```diff
@@ -335,6 +337,7 @@ def reverse_runtime_dependencies
     reverse_dependencies.where("d.scope ='runtime'")
   end
 
+  # TODO: rename this to clarify owner-defined mfa requirement vs policy mandated mfa requirement
```
I quickly grepped through the repo to see if there are any `TODO` comments. It looks like there's not many, just a handful. I'm hesitant about adding a TODO in the RubyGems codebase. If we're meaning to tackle this in the near future, I'd recommend opening an issue instead of adding this TODO inline with the code. Thoughts?

bettymakes commented on test/unit/rubygem_test.rb:
```diff
@@ -895,6 +895,28 @@ class RubygemTest < ActiveSupport::TestCase
+        rubygem_id: rubygem.id
+      )
+
+      refute_empty Rubygem.mfa_required_from_downloads
```
I'd add an assertion to verify that `rubygem` is within the collection of `Rubygem.mfa_required_from_downloads`.

kevinlinxc replied:
Good idea, I'm going to remove the original since it should be redundant when I add the includes? check

bettymakes commented on test/unit/user_test.rb:
```diff
@@ -370,7 +370,7 @@ class UserTest < ActiveSupport::TestCase
     end
 
-    context "recommend mfa" do
+    context "mfa recommend or require from downloads" do
```
NIT: I think it'd be better to use the same tense as everywhere else. In other words:

```
context "mfa recommended or required from downloads" do
```

Conversely, this could work too:
```
context "recommend or require mfa from downloads" do
```

(PS this "from downloads" wording _really_ discombobulates my understanding for some reason)

bettymakes commented on app/models/rubygem.rb:
```diff
@@ -34,8 +34,10 @@ class Rubygem < ApplicationRecord
+  MFA_REQUIRED_THRESHOLD = 180_000_000
 
   scope :mfa_recommended, -> { joins(:gem_download).where("gem_downloads.count > ?", MFA_RECOMMENDED_THRESHOLD) }
+  scope :mfa_required_from_downloads, -> { joins(:gem_download).where("gem_downloads.count > ?", MFA_REQUIRED_THRESHOLD) }
```
I'm not sure if I love this name yet 😅. What alternatives did y'all consider? I found it a bit confusing to wrap my mind around initially. 

Here are some ideas, just spitballing (they're not necessarily amazing ideas either lol):
* `mfa_required_per_download_threshold`
* `mfa_required_per_threshold`
* `mfa_required_as_per_download_threshold`
* `mfa_required_as_per_threshold`
* `mfa_required_from_download_threshold`
* `mfa_required_threshold_reached`
* `mfa_required_reached_threshold`
* `mfa_required_threshold_cleared`
* `mfa_required_cleared_threshold`

There's something about "from_downloads" that throws me off, it feels like it's implying something else.

kevinlinxc replied:
I suggest we just change mfa_required (from policy) to something else and then use mfa_required here so the scopes have the same naming scheme

bettymakes reviewed (commented):
Left a few notes, food for thought 😊. Overall, it looks like it sets up the groundwork well for the next set of work.

That said, it'd be great if the PR description was more detailed and provided insight into what these methods are in preparation for (especially since the methods aren't actually being used and are simply being defined in this PR).

jenshenny commented:
> 🤔 It is hard to do review when those methods are not used yet. Are they going to be part of some next PR as well?

@simi good point, sorry for not being transparent. I believe the plan is to do something similar with what we did with the warnings.

1. Create a PR with the mfa_required methods
2. Create a PR to use the mfa_required methods for the UI (this time redirecting to MFA pages on authenticated pages like https://github.com/Shopify/rubygems.org/pull/10/files) behind a feature cookie
3. Create a PR to use mfa_required methods for the CLI and merge on release (blocking gem push, yank, etc like https://github.com/Shopify/rubygems.org/pull/9/commits/6958f7caf26d146b72c9a25239ad99f19e0875e5)
4. Create a PR to remove the required cookie on release.

Maybe it makes more sense to have 1+2 in this PR as they can be merged before launch. The original rationale to have 1 separate is that we can do 2 and 3 in parallel.

kevinlinxc commented:
Implemented all of @bettymakes feedback and also changed:

rubygem.`mfa_required_from_downloads` scope to `mfa_required`
rubygem.`mfa_required? `method to `metadata_mfa_required?`
version.`rubygems_mfa_required?` method to `rubygems_metadata_mfa_required?` (Maybe too long). 

I think adding the distinction between policy and metadata mfa_required here makes sense, as Rubygem.mfa_required? and User.mfa_required? are now related, as you would expect, compared to before this refactor.

sonalkr132 commented on app/models/concerns/user_multifactor_methods.rb:
```diff
@@ -61,11 +69,17 @@ def strong_mfa_level?
 
     def mfa_recommended?
-      return false if strong_mfa_level?
+      return false if strong_mfa_level? || mfa_required?
```
We can remove `strong_mfa_level?` from here.

kevinlinxc commented:
Removed the strong_mfa_level? check, and rebase/squashed
